### PR TITLE
Fix setting this in Python when __slots__ are used in a class

### PR DESCRIPTION
**Pull request by hmaarrfk**
Closes: https://github.com/swig/swig/issues/1673

### CODE REVIEW

**Comment by hmaarrfk** [comment]
<details> <summary> Addition to the example </summary>

```diff
$ git diff -u class
diff --git a/Examples/python/class/example.i b/Examples/python/class/example.i
index fbdf7249f..7a62b6574 100644
--- a/Examples/python/class/example.i
+++ b/Examples/python/class/example.i
@@ -7,3 +7,9 @@
 
 /* Let's just grab the original header file here */
 %include "example.h"
+
+%extend Shape{
+    %pythoncode %{
+        __slots__ = ["this"]
+    %}
+};
```
</details>

**Comment by hmaarrfk** [comment]
Before:
```
✔ ~/git/swig/Examples/python/class_with_slots [:b0b69d65c|✔] 
00:41 $ make
make -f ../../Makefile SRCDIR='' CXXSRCS='example.cxx' \
SWIG_LIB_DIR='../../../Lib' SWIGEXE='../../../swig' \
TARGET='example' INTERFACE='example.i' python_cpp
make[1]: Entering directory '/home/mark/git/swig/Examples/python/class_with_slots'
env SWIG_LIB=../../../Lib  ../../../swig -python  -c++  -o example_wrap.cxx example.i
g++ -c -fpic -I. -I/include/boost-0  example_wrap.cxx  example.cxx   -I/usr/include/python2.7 -I/usr/lib/python2.7/config
g++ -shared  -I/include/boost-0    example.o example_wrap.o    -o _example.so
make[1]: Leaving directory '/home/mark/git/swig/Examples/python/class_with_slots'
make -f ../../Makefile SRCDIR='' python_run
make[1]: Entering directory '/home/mark/git/swig/Examples/python/class_with_slots'
env PYTHONPATH=$PWD  python  runme.py 
Creating some objects:
make[1]: *** [../../Makefile:395: python_run] Segmentation fault (core dumped)
make[1]: Leaving directory '/home/mark/git/swig/Examples/python/class_with_slots'
make: *** [Makefile:10: check] Error 2
```

After
```
00:41 $ git checkout fix_setthis_python 
Previous HEAD position was b0b69d65c Add an example with the slots
Switched to branch 'fix_setthis_python'
Your branch is up to date with 'origin/fix_setthis_python'.
✔ ~/git/swig/Examples/python/class_with_slots [fix_setthis_python|✔] 
00:41 $ make
make -f ../../Makefile SRCDIR='' CXXSRCS='example.cxx' \
SWIG_LIB_DIR='../../../Lib' SWIGEXE='../../../swig' \
TARGET='example' INTERFACE='example.i' python_cpp
make[1]: Entering directory '/home/mark/git/swig/Examples/python/class_with_slots'
env SWIG_LIB=../../../Lib  ../../../swig -python  -c++  -o example_wrap.cxx example.i
g++ -c -fpic -I. -I/include/boost-0  example_wrap.cxx  example.cxx   -I/usr/include/python2.7 -I/usr/lib/python2.7/config
g++ -shared  -I/include/boost-0    example.o example_wrap.o    -o _example.so
make[1]: Leaving directory '/home/mark/git/swig/Examples/python/class_with_slots'
make -f ../../Makefile SRCDIR='' python_run
make[1]: Entering directory '/home/mark/git/swig/Examples/python/class_with_slots'
env PYTHONPATH=$PWD  python  runme.py 
Creating some objects:
    Created circle <example.Circle; proxy of <Swig Object of type 'Circle *' at 0x7f2e8095cd50> >
    Created square <example.Square; proxy of <Swig Object of type 'Square *' at 0x7f2e8095cc90> >

A total of 2 shapes were created

Here is their current position:
    Circle = (20.000000, 30.000000)
    Square = (-10.000000, 5.000000)

Here are some properties of the shapes:
    <example.Circle; proxy of <Swig Object of type 'Circle *' at 0x7f2e8095cd50> >
        area      =  314.159265359
        perimeter =  62.8318530718
    <example.Square; proxy of <Swig Object of type 'Square *' at 0x7f2e8095cc90> >
        area      =  100.0
        perimeter =  40.0

Guess I'll clean up now
0 shapes remain
Goodbye
make[1]: Leaving directory '/home/mark/git/swig/Examples/python/class_with_slots'
```

File: Examples/python/class_with_slots/example.i [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/* File : example.i */
+%module example
+
+%{
+    #include "example.h"
+%}
+
+
+%extend Shape{
```
Comment (hmaarrfk) [inline Examples/python/class_with_slots/example.i]: the only difference between the two tests is the addition of the slots. 

Most of what slots do for me is they help me avoid typos once the API get mostly stable.

**Comment by hmaarrfk** [comment]
bump on comments regarding this issue?

**Comment by vadz** [comment]
I don't think it's worth adding a whole new example minimally differing from the existing one just for this, it really ought to be a unit test.

Concerning the changes to the code itself, it looks like the original author of this code considered the version directly manipulating the dicts to be faster than the one using `PyObject_GetAttrString()`. If this is no longer the case, we should indeed use `PyObject_SetAttrString()` if only because it's much simpler, but it would be nice to bench mark this to check that it's not going to result in any important performance regression.

**Comment by hmaarrfk** [comment]
Good points, can you point me to where I should add a unit test?

How do you benchmark different strategies?

**Comment by vadz** [comment]
Python unit tests are under `Examples/test-suite/python` directory, with each `foo_runme.py` being used to run the test using `foo.i` (in the parent directory) when you do `make -C Examples/test-suite/python foo.cpptest` or when you run `make check-python-test-suite` to run the full test suite.

If you can find some remotely related test it would probably be best to just add `__slots__` assignments to it, but otherwise you can also create a new `python_slots.i` test and `python/python_slots_runme.py` and add `python_slots` to `CPP_TEST_CASES` in `Examples/test-suite/python/Makefile.in`.

As for benchmarking, I don't have any particularly smart way to test it, but the test suite could be reused for this too, i.e. simply run `time make check-python-test-suite` (probably at least a few times, to get a reasonable average/discard outliers) before and after your changes and check if there is any noticeable difference.

Good luck!

**Pull request closed without merging**